File: src/services/scraper.ts
import axios, { AxiosError } from 'axios';
import * as cheerio from 'cheerio';
import { LiveData } from '../schemas/itinerarySchemas';

// A simple list of user agents to rotate through.
// In a production app, this should come from a configuration file.
const USER_AGENTS = [
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36',
  'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36',
  'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/109.0.0.0 Safari/537.36',
];

const getRandomUserAgent = (): string => {
  return USER_AGENTS[Math.floor(Math.random() * USER_AGENTS.length)];
};

/**
 * Extracts data from embedded application/ld+json scripts. This is the primary, most reliable method.
 * @param $ A cheerio instance representing the document.
 * @returns A partial LiveData object containing any found data.
 */
const _extractFromLdJson = ($: cheerio.CheerioAPI): Partial<Omit<LiveData, 'scraped_at' | 'error'>> => {
  const extractedData: Partial<Omit<LiveData, 'scraped_at' | 'error'>> = {};

  $('script[type="application/ld+json"]').each((_, element) => {
    try {
      const scriptContent = $(element).html();
      if (!scriptContent) return;

      const data = JSON.parse(scriptContent);
      const graph = data['@graph'] || [data];

      for (const item of graph) {
        if (typeof item !== 'object' || item === null) continue;

        const type = item['@type'] || '';
        const isPlace = ['Place', 'LocalBusiness', 'Restaurant', 'TouristAttraction', 'Museum'].some(placeType => type.includes(placeType));

        if (isPlace) {
          if (item.openingHours && !extractedData.operating_hours) {
            extractedData.operating_hours = Array.isArray(item.openingHours) ? item.openingHours.join(', ') : String(item.openingHours);
          }
          if (item.aggregateRating?.ratingValue && !extractedData.rating) {
            extractedData.rating = String(item.aggregateRating.ratingValue);
          }
          if (item.url && !extractedData.website_url) {
            extractedData.website_url = item.url;
          }
        }
      }
    } catch (e) {
      console.error('Failed to parse a JSON-LD script:', e);
    }
  });

  return extractedData;
};

/**
 * Scrapes a single URL for place information using axios and cheerio.
 * It prioritizes structured JSON-LD data for reliability.
 * @param url The URL to scrape.
 * @returns A promise that resolves to a LiveData object.
 */
export const scrapeUrl = async (url: string): Promise<LiveData> => {
  const scraped_at = new Date().toISOString();

  try {
    const response = await axios.get(url, {
      headers: { 'User-Agent': getRandomUserAgent() },
      timeout: 15000, // 15-second timeout for resilience
    });

    const html = response.data;
    const $ = cheerio.load(html);

    const scrapedData = _extractFromLdJson($);

    // Fallback for website URL if not found in structured data.
    if (!scrapedData.website_url) {
      scrapedData.website_url = url;
    }

    return {
      operating_hours: scrapedData.operating_hours ?? null,
      rating: scrapedData.rating ?? null,
      website_url: scrapedData.website_url ?? null,
      scraped_at,
      error: null,
    };
  } catch (error) {
    let errorMessage = `An unknown error occurred while processing ${url}.`;
    if (axios.isAxiosError(error)) {
      errorMessage = `Failed to fetch ${url}: ${error.message}`;
    } else if (error instanceof Error) {
      errorMessage = `An error occurred during parsing for ${url}: ${error.message}`;
    }

    console.error(errorMessage);
    return {
      scraped_at,
      error: errorMessage,
      operating_hours: null,
      rating: null,
      website_url: null,
    };
  }
};

File: orion_service/scraper.py
import asyncio
import json
import logging
from datetime import datetime, timezone

import aiohttp
from bs4 import BeautifulSoup

from orion_service import config, schemas

# Configure logging for the module
logging.basicConfig(level=logging.INFO, format='%(asctime)s - [%(levelname)s] - %(message)s')

def _extract_from_ld_json(soup: BeautifulSoup) -> dict:
    """
    Finds and parses application/ld+json scripts in the page to extract structured data.
    It checks for common schema.org types related to places.
    """
    extracted_data = {}
    try:
        ld_json_scripts = soup.find_all('script', type='application/ld+json')
        for script in ld_json_scripts:
            if not script.string:
                continue

            data = json.loads(script.string)
            # A page can have a single JSON-LD object or a @graph list of objects
            graph = data.get('@graph', [data])
            for item in graph:
                # Ensure item is a dictionary before processing
                if not isinstance(item, dict):
                    continue

                item_type = item.get('@type', '')
                is_place = isinstance(item_type, str) and item_type in [
                    'Place', 'LocalBusiness', 'Restaurant', 'TouristAttraction', 'Museum'
                ]
                if not is_place:
                    continue

                # Extract data points, only if not already found
                if 'openingHours' in item and not extracted_data.get('operating_hours'):
                    hours = item['openingHours']
                    extracted_data['operating_hours'] = ", ".join(hours) if isinstance(hours, list) else str(hours)

                if 'aggregateRating' in item and not extracted_data.get('rating'):
                    rating_info = item.get('aggregateRating', {})
                    if 'ratingValue' in rating_info:
                        extracted_data['rating'] = str(rating_info['ratingValue'])

                if 'url' in item and not extracted_data.get('website_url'):
                    extracted_data['website_url'] = item['url']

                # If we have all key data points, we can stop parsing this graph
                if 'operating_hours' in extracted_data and 'rating' in extracted_data:
                    break
    except (json.JSONDecodeError, TypeError, AttributeError) as e:
        logging.warning(f"Could not parse LD+JSON data: {e}")

    return extracted_data

async def _fetch_html(session: aiohttp.ClientSession, url: str) -> str:
    """
    Fetches HTML content from a URL.
    This function will raise exceptions on failure, to be handled by the caller.
    """
    headers = {'User-Agent': config.get_random_user_agent()}
    timeout = aiohttp.ClientTimeout(total=config.REQUEST_TIMEOUT_SECONDS)
    async with session.get(url, headers=headers, timeout=timeout, allow_redirects=True) as response:
        response.raise_for_status()
        return await response.text()

async def scrape_url(url: str, session: aiohttp.ClientSession) -> schemas.LiveData:
    """
    Scrapes a single URL for place information.

    This function attempts to find structured (JSON-LD) data first, then
    falls back to simple heuristics if needed. It handles network errors gracefully.
    """
    start_time = datetime.now(timezone.utc)
    try:
        html = await _fetch_html(session, url)
        soup = BeautifulSoup(html, 'html.parser')

        # Primary Strategy: Parse structured JSON-LD data
        scraped_data = _extract_from_ld_json(soup)
        logging.info(f"For URL {url}, extracted from JSON-LD: {scraped_data}")

        # Fallback for website_url
        if not scraped_data.get('website_url'):
            scraped_data['website_url'] = url

        return schemas.LiveData(
            operating_hours=scraped_data.get('operating_hours'),
            rating=scraped_data.get('rating'),
            website_url=scraped_data.get('website_url'),
            scraped_at=start_time,
        )

    except (aiohttp.ClientError, asyncio.TimeoutError) as e:
        logging.warning(f"Network error while fetching {url}: {e}")
        return schemas.LiveData(
            scraped_at=start_time,
            error=f"Failed to fetch URL due to network error: {e}",
        )
    except Exception as e:
        logging.error(f"An unexpected error occurred while scraping {url}: {e}", exc_info=True)
        return schemas.LiveData(
            scraped_at=start_time,
            error=f"An unexpected error occurred while processing {url}: {e}",
        )
